src/functions/avl-consumer-subscriber/index.ts

```diff
@@ -0,0 +1,139 @@
+import { XMLParser } from "fast-xml-parser";
+import { ZodError, z } from "zod";
+
+const requestHeadersSchema = z.object({
+    userId: createStringLengthValidation("userId header"),
+});
+
+const requestParamsSchema = z.preprocess(
+    Object,
+    z.object({
+        subscriptionId: createSubscriptionIdArrayValidation("subscriptionId"),
+    }),
+);
+
+const requestBodySchema = z.string({
+    required_error: "Body is required",
+    invalid_type_error: "Body must be a string",
+});
+
+const parseXml = (xml: string) => {
+    const parser = new XMLParser({
+        allowBooleanAttributes: true,
+        ignoreAttributes: true,
+        parseTagValue: false,
+    });
+
+    const parsedXml = parser.parse(xml);
+    const parsedJson = avlSubscriptionRequestSchema.safeParse(parsedXml);
+
+    if (!parsedJson.success) {
+        throw new InvalidXmlError();
+    }
+
+    return parsedJson.data;
+};
+
+export const handler: APIGatewayProxyHandler = async (event, context) => {
+    withLambdaRequestTracker(event ?? {}, context ?? {});
+
+    try {
+        const {
+            AVL_CONSUMER_SUBSCRIPTION_TABLE_NAME: avlConsumerSubscriptionTableName,
+            AVL_PRODUCER_SUBSCRIPTION_TABLE_NAME: avlProducerSubscriptionTableName,
+        } = process.env;
+
+        if (!avlConsumerSubscriptionTableName || !avlProducerSubscriptionTableName) {
+            throw new Error(
+                "Missing env vars - AVL_CONSUMER_SUBSCRIPTION_TABLE_NAME and AVL_PRODUCER_SUBSCRIPTION_TABLE_NAME must be set",
+            );
+        }
+
+        const { userId } = requestHeadersSchema.parse(event.headers);
+        const requestParams = requestParamsSchema.parse(event.queryStringParameters);
+        const producerSubscriptionIds = requestParams.subscriptionId;
+
+        const body = requestBodySchema.parse(event.body);
+        const xml = parseXml(body);
+        const subscriptionRequest = xml.Siri.SubscriptionRequest;
+        const subscriptionId = subscriptionRequest.VehicleMonitoringSubscriptionRequest.SubscriptionIdentifier;
+
+        const isActiveSubscription = await isActiveAvlConsumerSubscription(
+            avlConsumerSubscriptionTableName,
+            subscriptionId,
+            userId,
+        );
+
+        if (isActiveSubscription) {
+            return createConflictErrorResponse("Consumer subscription ID already active");
+        }
+
+        const producerSubscriptions = await getAvlSubscriptions(avlProducerSubscriptionTableName);
+
+        for (const producerSubscriptionId of producerSubscriptionIds.split(",")) {
+            const subscription = producerSubscriptions.find(({ PK }) => PK === producerSubscriptionId);
+
+            if (!subscription || subscription.status === "inactive") {
+                return createNotFoundErrorResponse(`Producer subscription ID not found: ${producerSubscriptionId}`);
+            }
+        }
+
+        const consumerSubscription: AvlConsumerSubscription = {
+            PK: randomUUID(),
```

Comment:
In the case where there's an existing subscription in an error state that a user is resubscribing to, we probably want to keep the same PK here